wave_vision/include/wave/vision/tracker/tracker.hpp

```diff
@@ -0,0 +1,171 @@
+        return this->measurements.size();
+    }
+
+    /** The id of the feature */
+    size_t id;
+
+    /** All measurements of the feature */
+    std::vector<LandmarkMeasurement<int>> measurements;
+
+    /** The first image the feature is seen in */
+    size_t first_image;
+
+    /** The last image the feature is seen in */
+    size_t last_image;
+};
+
+template <typename TDetector, typename TDescriptor, typename TMatcher>
+class Tracker {
+ public:
+    /** Default constructor
+     *
+     * @param detector detector object (FAST, ORB, etc...)
+     * @param descriptor descriptor object (BRISK, ORB, etc...)
+     * @param matcher matcher object (BruteForceMatcher, FLANN)
+     */
+    Tracker(TDetector detector, TDescriptor descriptor, TMatcher matcher)
+        : detector(detector), descriptor(descriptor), matcher(matcher) {}
+
+    /** Destructor */
+    ~Tracker() = default;
+
+    /** Get the tracks of all features in the requested image from the sequence.
+     *
+     * @param img_num the number of the image to obtain tracks from
+     * @return tracks corresponding to all detected landmarks in the image.
+     */
+    std::vector<FeatureTrack> getTracks(const size_t &img_num) const;
+
+    /** Track features within an image (presumably the next in a sequence).
+     *
+     * @param image the image to add.
+     */
+    void addImage(const cv::Mat &image);
+
+    /** Draw tracks for the requested image.
+     *
+     * @param img_num the number of the image within the sequence
+     * @param image the image to draw the tracks on
+     * @return the image with the tracks illustrated as arrows.
+     */
+    cv::Mat drawTracks(const size_t &img_num, const cv::Mat &image) const;
+
+    /** Offline feature tracking, using list of images already loaded.
+     *
+     * @param image_sequence the sequence of images to analyze.
+     * @return the vector of FeatureTracks in each image.
+     */
+    std::vector<std::vector<FeatureTrack>> offlineTracker(
+      const std::vector<cv::Mat> &image_sequence);
+
+ private:
+    /** Generate a new ID for each newly detected feature.
+     *
+     * @return the assigned ID.
+     */
+    size_t generateFeatureID() const {
+        static size_t id = 0;
+        return id++;
+    }
+
+    /** Detect features and compute descriptors.
+     *
+     * Detects features and computes descriptors using the templated detector
+     * and descriptor.
+     *
+     * @param image
+     * @param keypoints
+     * @param descriptor
+     */
+    void detectAndCompute(const cv::Mat &image,
+                          std::vector<cv::KeyPoint> &keypoints,
+                          cv::Mat &descriptor);
+
+    /** Register the current time with the current img_count
+     *
+     * @param current_time the time at which this image was received
+     */
+    void timestampImage(
+      const std::chrono::steady_clock::time_point &current_time) {
+        this->img_times[this->img_count] = current_time;
```

Comment:
Ah turns out they're not _quite_ equal since C++ is zero indexed. It's still unnecessary though, and was more confusing.